src/client/fuse_ll.cc

```diff
@@ -651,6 +651,44 @@ static void fuse_ll_ioctl(fuse_req_t req, fuse_ino_t ino, int cmd, void *arg, st
       fuse_reply_ioctl(req, 0, &l, sizeof(struct ceph_ioctl_layout));
     }
     break;
+    #if defined(__linux__)
+    case FS_IOC_GETFLAGS:{
+       int attr = 0;
+       Fh *fh = (Fh*)fi->fh;
+       const struct fuse_ctx *ctx = fuse_req_ctx(req);
+       UserPerm perms(ctx->uid, ctx->gid);
+       get_fuse_groups(perms, req);
+       int ret = cfuse->client->ll_get_flags(fh, perms);
+       if (!ret) {
+         attr |= FS_IMMUTABLE_FL;
+       }
+       
+       fuse_reply_ioctl(req, 0, &attr, sizeof(attr));
+    }
+    break;
+    case FS_IOC_SETFLAGS: {
+      auto flags = (int *)in_buf;
+      
+      if (*flags & (~FS_IMMUTABLE_FL)) {
+        fuse_reply_err(req, EOPNOTSUPP);
+        return;
+      }
```

Comment:
I'm a little concerned with this. Suppose I accidentally set this on the wrong file. How do I fix it?

Reply:
I think may be only wait for file protection time expires 。any suggestion？

Reply:
I don't think waiting will be acceptable -- we will need to be able to fix this in some fashion, as we can't just throw away the filesystem and start over (as you would with a CD-R or something).

Maybe this could only be allowed from certain clients using a new cephfs client capability flag, and those clients would be able to undo this as well?

http://docs.ceph.com/docs/master/cephfs/client-auth/

Reply:
OK ! Already add a new cap flag "i" and "chattr -i"" to expired a file which is in retained state.